### ISSUE TITLE
Interoperability suggestions

### ISSUE THREAD

**Issue by bochoven** [UniversityPCCare/gsx-rest-php#1]
Dear dev,

Some people are looking eagerly at your efforts so they can reinstate the collection of warranty info to be used in [munkireport-php](https://github.com/munkireport/munkireport-php). To make things easier, maybe also for you, I would like to give some suggestions:

- Use composer and [packagist](https://packagist.org/) to embed your library. It makes life easier for people relying on your library. Packagist will automatically version the code when you release it.
-   Allow for passing in the configuration variables, make the ini parser a fallback mechanism if the variables aren’t passed in. This will help programs to manage their own config mechanism.
- Make storing sessions in a database optional. I don’t think the original library had persistence for sessions and while it may speed up queries (does it?) it complicates inclusion - prevent db name collisions, db upgrades etc.
- Lastly, if you are going to use Composer, you might as well take advantage of the excellent [monolog](https://packagist.org/packages/monolog/monolog) logging library

**Comment by upccjesse** [comment]
Thanks for the suggestions. This is my first public project and so I'm working to make it as widely accessible as possible. I wrote it primarily with the intention of integrating it into my own existing workflow, which is PHP+MySQL heavy, so it "just works" for me. I considered including multiple persistence options similar to other programs I've seen but it wasn't a priority when I created the project. To address your points:

- Composer + packagist: I will look into this.
- Passing in configuration variables: That can reasonably be done.
- The original library did not need to persist anything because the SOAP API did not require persistence of any kind. Authentication for the SOAP API was based on whitelisted IP + valid Apple-signed certificate + email address with "web access" enabled for GSX. The new REST API additionally requires (per user) a one-time-use activation token, and then an authorization token that will be used for up to 12 hours after creation or 30 minutes after last use, or until a call is made to the `/authenticate/end-session` endpoint, whichever comes first. Because of this, there needs to be some persistence to store the authorization token to be used over multiple requests. For me, it was a no-brainer to use MySQL as we already have that instance setup.
  - As for DB name collisions, you can specify the database name in the `config.ini` file.
- My first thought was to depend on as few things as possible, which is why I didn't include any dependencies. If I do setup Composer + packagist, then I will implement monolog.

Thank you for your interest in the project.

**Comment by bochoven** [comment]
> Because of this, there needs to be some persistence to store the authorization token to be used over multiple requests.

Is it possible to request a new authorization token with each request? That way, you don’t need persistence

**Comment by upccjesse** [comment]
No, because to get a new authorization token you either need the one-time use activation token (which must be retrieved manually from Apple) or your most recent authorization token. The only way to not have persistence is to manually retrieve a new activation token for each contiguous batch of requests (ie, every time your PHP script runs) which is unfeasible. Apple also specifically warns about "unexpected behavior" if you call the `/authenticate/token` API too much.

To note: the only reason I store the one-time use activation token is so users can manually retrieve the activation token from Apple and store it (like, in a Settings page) to be used later.

**Comment by upccjesse** [comment]
Do you have any thoughts on best practice for passing in the variables? I'm thinking of three possible ways:

- Parameterize each option, ie `__construct($email, $shipTo, $soldTo, ...)`
- Pass in an array, ie `__construct(["email" => $email, ... ])`
- Create a separate class that will handle parameters and error checking, then pass that class in

Array seems like a good "go-to" option.

**Comment by upccjesse** [comment]
I implemented passing in `config.ini` options via array.

**Comment by Isaac-Duarte** [comment]
Hello Jesse!
I'm currently in the process of writing my own implementation of the GSX API in C# for my company. I'm currently in the process working with my IT Department on getting access to the REST Api. I've created a base implementation as I do not have any idea of endpoints and such. My question is for the one-time authentication code. How is that specifically obtained? I know you mentioned `users can manually retrieve the activation token from Apple` Is this a text message type thing like logging into GSX? Thank you! I am just wanting to get an Idea so I can start writing some code beforehand to speed up the process.

**Comment by upccjesse** [comment]
You'll need to work with the GSX Web Support team and they'll get you the links and information you need. They'll also set you up in the test/Dev environment first. They're pretty responsive during the working week.

**Comment by Isaac-Duarte** [comment]
Okay thanks man I appreciate your response!
